pkg/config/msp.go

```diff
@@ -262,3 +268,156 @@ func pemEncodeECDSAPrivateKey(priv *ecdsa.PrivateKey) ([]byte, error) {
+	}
+
+	org, err := getApplicationOrg(config, orgName)
+	if err != nil {
+		return err
+	}
+	fabricMSPConfig, err := getFabricMSPConfig(org)
+	if err != nil {
+		return fmt.Errorf("getting msp config: %v", err)
+	}
+
+	certBytes := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: rootCA.Raw})
+	if err != nil {
+		return err
+	}
+	fabricMSPConfig.RootCerts = append(fabricMSPConfig.RootCerts, certBytes)
+
+	err = addMSPConfigToOrg(org, fabricMSPConfig)
+	if err != nil {
+		return fmt.Errorf("adding msp config to org: %v", err)
+	}
+
+	return nil
+}
+
+func getFabricMSPConfig(org *cb.ConfigGroup) (*mb.FabricMSPConfig, error) {
+	configValue, err := getOrgMSPValue(org)
+	if err != nil {
+		return nil, err
+	}
+
+	mspConfig := &mb.MSPConfig{}
+	err = proto.Unmarshal(configValue.Value, mspConfig)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling mspConfig: %v", err)
+	}
+
+	fabricMSPConfig := &mb.FabricMSPConfig{}
+	err = proto.Unmarshal(mspConfig.Config, fabricMSPConfig)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling mspConfig: %v", err)
+	}
+
+	return fabricMSPConfig, nil
+}
+
+func addMSPConfigToOrg(org *cb.ConfigGroup, fabricMSPConfig *mb.FabricMSPConfig) error {
+	configValue, err := getOrgMSPValue(org)
+	if err != nil {
+		return err
+	}
+
+	mspConfig := &mb.MSPConfig{}
+	err = proto.Unmarshal(configValue.Value, mspConfig)
+	if err != nil {
+		return fmt.Errorf("unmarshalling mspConfig: %v", err)
+	}
+
+	serializedMSPConfig, err := proto.Marshal(fabricMSPConfig)
+	if err != nil {
+		return fmt.Errorf("marshalling updated mspConfig: %v", err)
+	}
+
+	mspConfig.Config = serializedMSPConfig
+
+	err = addValue(org, mspValue(mspConfig), AdminsPolicyKey)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getOrgMSPValue(org *cb.ConfigGroup) (*cb.ConfigValue, error) {
+	configValue, ok := org.Values[MSPKey]
+	if !ok {
+		return nil, errors.New("org doesn't have MSP value")
+	}
+	return configValue, nil
+}
+
+// RevokeCertificateFromMSP takes an x509 certificate and adds it to the
+// revocation list for the specified application org MSP. It signs the CRL
+// using the specified crypto.Signer.
+func RevokeCertificateFromMSP(cert *x509.Certificate, orgName string, signer crypto.Signer, config *cb.Config) error {
+	org, err := getApplicationOrg(config, orgName)
+	if err != nil {
+		return err
+	}
+
+	fabricMSPConfig, err := getFabricMSPConfig(org)
+	if err != nil {
+		return fmt.Errorf("getting msp config: %v", err)
+	}
+
+	// TODO validate that this certificate was issued by this MSP
+
+	if len(fabricMSPConfig.RevocationList) == 0 {
+		// TODO create a CRL here if one doesn't exist?
+		return errors.New("no CRL in RevocationList")
+	}
+
+	// just parse the first CRL and add it there for now
+	crl, err := x509.ParseCRL(fabricMSPConfig.RevocationList[0])
+	if err != nil {
+		return fmt.Errorf("parsing CRL: %v", err)
+	}
+
+	crl.TBSCertList.RevokedCertificates = append(crl.TBSCertList.RevokedCertificates,
```

Comment:
This troubles me. I would fully expect us to use the CA certificate instance to create a new CRL and the associated private key to sign it.

Basically, everything from 388 onward is part of `CreateCRL` in the standard library.

Reply:
Done